test/golden/generated_code.d

```diff
@@ -9,26 +9,26 @@ enum protocVersion = 3005000;
 
 class TestAllTypes
 {
-    @Proto(1) int optionalInt32 = protoDefaultValue!int;
-    @Proto(2) long optionalInt64 = protoDefaultValue!long;
-    @Proto(3) uint optionalUint32 = protoDefaultValue!uint;
-    @Proto(4) ulong optionalUint64 = protoDefaultValue!ulong;
-    @Proto(5, Wire.zigzag) int optionalSint32 = protoDefaultValue!int;
-    @Proto(6, Wire.zigzag) long optionalSint64 = protoDefaultValue!long;
-    @Proto(7, Wire.fixed) uint optionalFixed32 = protoDefaultValue!uint;
-    @Proto(8, Wire.fixed) ulong optionalFixed64 = protoDefaultValue!ulong;
-    @Proto(9, Wire.fixed) int optionalSfixed32 = protoDefaultValue!int;
-    @Proto(10, Wire.fixed) long optionalSfixed64 = protoDefaultValue!long;
-    @Proto(11) float optionalFloat = protoDefaultValue!float;
-    @Proto(12) double optionalDouble = protoDefaultValue!double;
-    @Proto(13) bool optionalBool = protoDefaultValue!bool;
-    @Proto(14) string optionalString = protoDefaultValue!string;
-    @Proto(15) bytes optionalBytes = protoDefaultValue!bytes;
-    @Proto(18) NestedMessage optionalNestedMessage = protoDefaultValue!NestedMessage;
-    @Proto(19) ForeignMessage optionalForeignMessage = protoDefaultValue!ForeignMessage;
-    @Proto(21) NestedEnum optionalNestedEnum = protoDefaultValue!NestedEnum;
-    @Proto(22) ForeignEnum optionalForeignEnum = protoDefaultValue!ForeignEnum;
-    @Proto(27) TestAllTypes recursiveMessage = protoDefaultValue!TestAllTypes;
+    @Proto(1) int optionalInt32 = protoDefaultValue!(int);
+    @Proto(2) long optionalInt64 = protoDefaultValue!(long);
+    @Proto(3) uint optionalUint32 = protoDefaultValue!(uint);
+    @Proto(4) ulong optionalUint64 = protoDefaultValue!(ulong);
+    @Proto(5, Wire.zigzag) int optionalSint32 = protoDefaultValue!(int);
+    @Proto(6, Wire.zigzag) long optionalSint64 = protoDefaultValue!(long);
+    @Proto(7, Wire.fixed) uint optionalFixed32 = protoDefaultValue!(uint);
+    @Proto(8, Wire.fixed) ulong optionalFixed64 = protoDefaultValue!(ulong);
+    @Proto(9, Wire.fixed) int optionalSfixed32 = protoDefaultValue!(int);
+    @Proto(10, Wire.fixed) long optionalSfixed64 = protoDefaultValue!(long);
+    @Proto(11) float optionalFloat = protoDefaultValue!(float);
+    @Proto(12) double optionalDouble = protoDefaultValue!(double);
+    @Proto(13) bool optionalBool = protoDefaultValue!(bool);
+    @Proto(14) string optionalString = protoDefaultValue!(string);
+    @Proto(15) bytes optionalBytes = protoDefaultValue!(bytes);
+    @Proto(18) TestAllTypes.NestedMessage optionalNestedMessage = protoDefaultValue!(TestAllTypes.NestedMessage);
```

Comment:
`NestedMessage` shouldn't be prefixed with the parent class. It shouldn't be different than `NestedEnum`.
Do I miss something?

Reply:
Done. The reason why I fully qualified types: it is the simplest implementation and valid in D. I agree that the simple generated code is more important than the simple code generator.

During adding unittests, I found another bug in `oneofAccessors`. I finally figured out that stringof ignores parent types, and std.traits.fullyQualifiedName is recommended for code generation.
https://dlang.org/spec/property.html#stringof